scrapy/downloadermiddlewares/httpproxy.py

```diff
@@ -26,30 +52,13 @@ def from_crawler(cls, crawler):
         auth_encoding = crawler.settings.get('HTTPPROXY_AUTH_ENCODING')
         return cls(auth_encoding)
 
-    def _basic_auth_header(self, username, password):
-        user_pass = to_bytes(
-            '%s:%s' % (unquote(username), unquote(password)),
-            encoding=self.auth_encoding)
-        return base64.b64encode(user_pass).strip()
-
-    def _get_proxy(self, url, orig_type):
-        proxy_type, user, password, hostport = _parse_proxy(url)
-        proxy_url = urlunparse((proxy_type or orig_type, hostport, '', '', '', ''))
-
-        if user:
-            creds = self._basic_auth_header(user, password)
-        else:
-            creds = None
-
-        return creds, proxy_url
-
```

Comment:
Removing these methods is backwards-incompatible, because HttpProxyMiddleware subclasses which override or call these methods will stop working properly. I know they are private and undocumented, but they are called (though not overridden) in real-world HttpProxyMiddleware subclasses - I've found examples in a large codebase. What do you think about keeping them, but maybe just calling functions from there, and maybe after issuing a deprecation warning?

Reply:
Reasonable, I have kept these two methods now, with a deprecated warning.

The message said the method will be removed in the future release, but which version will be the future release? Is there a schedule for deprecation in scrapy?

Reply:
We don't have scheduled removal of deprecated features; we keep them for +1 release for sure, and then we can keep them for much longer, depending on whether they cause maintenance issues. There was a bit of discussion about deprecation policy here: https://github.com/scrapy/scrapy/issues/2568.